keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,464 @@
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+// +normalize=undefined
+Containers []Container `json:containers,omitempty` `protobuf:"bytes,2,opt,name=containers"
+```
+
+For fields which carry no `omitempty`, we could allow more advanced normalization
+modes which replicate the Golang serialization behaviour. Tooling like openapi-gen
+and the CRD validation could verify that the normalization specification matches 
+that of Golang.
+
+### Mutating admission chain
+
+Mutating admission chain would have the exact same effects as user changes,
+meaning that they wouldn't be able to change an object after creation. This is
+very similar to what is done today since validation for updates is run AFTER all
+mutations.
+
+### Where does this happen
+
+This process is meant to happen right before the update validation and after
+mutating, but before validating webhooks, and only run on updates. This will allow us to
+keep the exact same behavior while removing the validation code that checks the
+immutability of fields.
+
+![Decoding steps which must apply defaults](20191001-crd-immutability-pipeline.png)
+
+### Risks and Mitigations
+
+- immutable metadata would break API machinery. We forbid the `x-kubernetes-immutable`
+  at the root of the object and inside `.metadata`. `kind` and `apiVersion` are
+  immutable implicitly. We might publish immutable though for some of these fields.
+
+## Design Details
+
+### Test Plan
+
+- exhaustive unit tests are added in apiextensions-apiserver for
+  - CRD validation
+    - for `x-kubernetes-immutable: true` and `x-kubernetes-immutable-keys: true` at the root and in `metadata`
+    - for `x-kubernetes-immutable-keys: true` for list-type `map` and forbidden otherwise
+    - for `x-kubernetes-immutable: true` on the list-keys of `x-kubernetes-immutable-keys: true` maps.
+    - for `x-kubernetes-immutable: true` and `x-kubernetes-immutable-keys: true` only for v1 CRDs, or during ratcheting updates.
+  - immutability checking with all variants of `x-kubernetes-immutable`, `x-kubernetes-immutable-keys` and `x-kubernetes-list-type`.
+- integration tests are added for
+  - creation, updates, patches and server-side-apply of partially immutable CRs 
+  - interaction of server-side-apply list-types and immutability
+  - OpenAPI publishing of the vendor extensions
+  - CRD updates of the immutability extensions and that the new immutability
+    schemas are followed.
+  - rejection of `x-kubernetes-immutable: true` and `x-kubernetes-immutable-keys: true` for non-v1 CRDs
+- e2e and conformance tests that
+  - immutability is followed during updates, patches and server-side-apply. 
+   
+### Graduation Criteria
+
+Because of the very limited risk of the additional immutability check and our
+experience with defaulting getting no additional soak time by the community
+we propose to start directly as beta.
```

Comment:
This results in potential data loss during HA API server upgrade or rollback (a 1.16 API server will drop the proposed fields). We need the n-1 API server to be capable of preserving the data in the field before we enable/allow it by default, which implies a 1.17 alpha and a 1.18 beta.

Reply:
I object to going straight to beta, but on the grounds that I object to adding the immutable-keys tag *at all* :)

Reply:
changed to go to alpha first.